irods/session.py

```diff
@@ -17,13 +19,49 @@
 from irods.exception import NetworkException
 from irods.password_obfuscation import decode
 from irods import NATIVE_AUTH_SCHEME, PAM_AUTH_SCHEME
+import threading
+import weakref
+
+_sessions = None
+_recycle_bin = weakref.WeakKeyDictionary()
+_sessions_lock = threading.Lock()
+
+GC_LIMIT = None # A positive integer indicates the threshold count of iRODSSession
+                # objects where cleanup() and garbage collection is performed.
+
+def _cleanup_remaining_sessions(do_all = True):
+    if do_all:
+        _sessions.update(_recycle_bin)
+        _recycle_bin.clear()
+    for ses in _sessions.copy():
+        ses.cleanup()  # internally modifies _sessions
+
+def _weakly_reference(ses, gc_limit=None):
+    global _sessions
+    try:
+        if _sessions is None:
+            with _sessions_lock:
```

Comment:
see if this or other lines in the try can throw. will _sessions be initialized for 100% certain when "finally" clause runs?

Reply:
Examined this.  It is not possible for the code above to throw (threading.Lock.acquire() can be interrupted by the OS, but this directly terminates the process instead of raising an exception). Therefore the finally clause following it will always have a valid _sessions dictionary.